Lib/fontTools/pens/hashPointPen.py

```diff
@@ -0,0 +1,57 @@
+# Modified from https://github.com/adobe-type-tools/psautohint/blob/08b346865710ed3c172f1eb581d6ef243b203f99/python/psautohint/ufoFont.py#L800-L838
+import hashlib
+
+from fontTools.pens.pointPen import AbstractPointPen
+from fontTools.pens.transformPen import TransformPointPen
+
+
+class HashPointPen(AbstractPointPen):
+    """
+    This pen can be used to check if a glyph's contents (outlines plus
+    components) has changed.
+
+    Components are decomposed, because that is what happens in TrueType
```

Comment:
IIUC that means a composite glyph with one component would hash the same as the base glyph it references. I find that a bit strange for a pen that is supposed to check if a glyph's contents have changed.

Reply:
Hm... the original hash pen from Adobe just added the base glyph name and transformation to the hash. But for detecting outline changes, especially for TrueType hinting, we need to make sure the component's outlines also haven't changed. On the other hand, a base glyph may have been renamed, so it may be the same base glyph even if the component references a different name.

Maybe adding a generic "component marker" (just to say "this path comes from a component" without specifying the base glyph name) to the hash and still adding the (transformed) component points verbatim would solve this?

Then again, only the base glyph's hinting may have changed, which may or may not break any additional TT instructions in the composite glyph, which can't be detected by this hashing mechanism and maybe is out of scope.

Perhaps it's best to be on the safe side and include the component names as well as the transformations in the hash, WDYT?

Reply:
how about recording the _original_ base glyphs' outlines followed by the respective transform? The name of the base glyph is irrelevant as you said, it's the content that this hash pen is recording.